pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ClassStub.java

```diff
@@ -134,6 +134,15 @@ protected void finishParse(boolean failed) {
                                   : HashTreePSet.empty();
             }
 
+            @Override
+            protected boolean canReenter() {
+                // We might call the parsing logic again in the same thread,
+                // e.g. in order to determine "annotAttributes", getDeclaredMethods() is called, which
+                // calls ensureParsed().
+                // Note: Other threads can't reenter, since our thread own the ParseLock monitor.
+                return true;
```

Comment:
That's a consequence that we set the status of the ParseLock to FULL later - finishParse() actually calls back into ensureParsed() and tries to do the parsing again. Since we are in the same thread, we already own the monitor and actually reenter the ParseLock - but the status is already "BEING_PARSED".

I think, for ClassStub it's ok to allow reenter without being picky. The only field, that is being finished in finishParse is annotAttributes - everything else is already in set.